packages/graph/hash_graph/lib/graph/src/store/query/filter.rs

```diff
@@ -189,17 +225,115 @@ impl<'q, T: QueryRecord> TryFrom<Expression> for Option<FilterExpression<'q, T>>
 #[derive(Debug, PartialEq, Deserialize)]
 #[serde(untagged)]
 pub enum Parameter<'q> {
+    Boolean(bool),
     Number(f64),
     Text(Cow<'q, str>),
-    Boolean(bool),
+    #[serde(skip)]
+    Uuid(Uuid),
+    #[serde(skip)]
+    SignedInteger(i64),
+}
+
+impl Parameter<'_> {
+    fn to_owned(&self) -> Parameter<'static> {
+        match self {
+            Parameter::Boolean(bool) => Parameter::Boolean(*bool),
+            Parameter::Number(number) => Parameter::Number(*number),
+            Parameter::Text(text) => Parameter::Text(Cow::Owned(text.to_string())),
+            Parameter::Uuid(uuid) => Parameter::Uuid(*uuid),
+            Parameter::SignedInteger(integer) => Parameter::SignedInteger(*integer),
+        }
+    }
+}
+
+#[derive(Debug)]
+#[must_use]
+pub struct ParameterConversionError {
+    actual: Parameter<'static>,
+    expected: ParameterType,
+}
+
+impl fmt::Display for ParameterConversionError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            fmt,
+            "could not convert `{}` to {}",
+            self.actual, self.expected
+        )
+    }
+}
+
+impl Context for ParameterConversionError {}
+
+impl Parameter<'_> {
+    #[expect(clippy::match_same_arms, reason = "Add validation of parameters")]
+    fn convert_to_parameter_type(
+        &mut self,
+        expected: ParameterType,
+    ) -> Result<(), Report<ParameterConversionError>> {
+        match (&mut *self, expected) {
+            (_, ParameterType::Any)
+            | (Parameter::Boolean(_), ParameterType::Boolean)
+            | (Parameter::Number(_), ParameterType::Number)
+            | (Parameter::Text(_), ParameterType::Text) => {
+                // no action needed, exact match
+            }
+            (Parameter::Text(_base_uri), ParameterType::BaseUri) => {
+                // TODO: validate base uri
```

Comment:
Do we perhaps want todo panics?

Reply:
An empty body here just ignores validation. Base URIs are stored as text in the database (and `BaseUri` does not implement `tokio_postgres::types::ToSql` to my knowledge) so `Text` is the right type here. We could just add some validation logic here as well. Alternatively, we could change the `ParameterType` for `BaseUri` and `VersionedUri` to `Text` instead.

Reply:
Makes sense, just concerned about the empty TODOs. When are we thinking about putting these in? Should we associate an asana task with them